### PR TITLE
optimize xml reading in TMVA (at least for CMS use case)

### PR DESCRIPTION
**Pull request by davidlange6**
Primary changes proposed
 - avoiding char to TString to char conversions
 - avoid redundant call to xmlEngine().hasattr(node,att) checking instead for null ptr from getattr.
 - Add specialization of void TMVA::Tools::ReadAttr( void* node, const char* attrname, T& value ) for float,int, short. [which seem to be the ones we use aside from bool, which now is the biggest contributor to ReadAttr* timing.

### CODE REVIEW

**Comment by vgvassilev** [comment]
Thanks for working on this! Could you address the clang-format and clang-tidy reports in travis?

**Comment by vgvassilev** [comment]
If you don't want to install clang-format you could download the diff and apply it locally from here: https://api.travis-ci.org/jobs/233691739/log.txt?deansi=true

**Comment by davidlange6** [comment]
ah - i still didn't get it right... learning

File: tmva/tmva/inc/TMVA/Tools.h [hunk 1]
```diff
@@ -283,9 +289,14 @@ namespace TMVA {
 
 template<typename T> void TMVA::Tools::ReadAttr( void* node, const char* attrname, T& value )
 {
-   TString val;
-   ReadAttr( node, attrname, val );
-   std::stringstream s(val.Data());
+   // read attribute from xml
+   const char *val = xmlengine().GetAttr(node, attrname);
+   if (val == 0) {
```
Comment (pcanal) [inline tmva/tmva/inc/TMVA/Tools.h]: Please use nullptr whenever applicable :)

File: tmva/tmva/src/Tools.cxx [hunk 2]
```diff
@@ -1790,3 +1790,39 @@ TH1* TMVA::Tools::GetCumulativeDist( TH1* h)
    }
    return cumulativeDist;
 }
+
+void TMVA::Tools::ReadAttr(void *node, const char *attrname, float &value)
+{
+   // read attribute from xml
+   const char *val = xmlengine().GetAttr(node, attrname);
+   if (val == NULL) {
```
Comment (pcanal) [inline tmva/tmva/src/Tools.cxx]: Please use nullptr whenever applicable :)

Comment (davidlange6) [inline tmva/tmva/src/Tools.cxx]: ok - changing..

File: tmva/tmva/inc/TMVA/Tools.h [hunk 3]
```diff
@@ -283,9 +289,14 @@ namespace TMVA {
...
+      const char *nodename = xmlengine().GetNodeName(node);
+      Log() << kFATAL << "Trying to read non-existing attribute '" << attrname << "' from xml node '" << nodename << "'"
+            << Endl;
+   }
```
Comment (pcanal) [inline tmva/tmva/inc/TMVA/Tools.h]: Didn't you mean to put an else here and/or a return before?

Comment (davidlange6) [inline tmva/tmva/inc/TMVA/Tools.h]: i had assumed kFATAL would take care of that:) - (the previous author did) - shall I add an else?

Comment (pcanal) [inline tmva/tmva/inc/TMVA/Tools.h]: My bad .. it does indeed ...

**Comment by vgvassilev** [comment]
@phsft-bot build!

**Comment by vgvassilev** [comment]
I think those failures are unrelated to this PR.

**Pull request merged**